gcc/rust/rust-diagnostics.h

```diff
@@ -51,15 +51,1034 @@
+  E0596,
+  E0597,
+  E0599,
+  E0600,
+  E0601,
+  E0602,
+  E0603,
+  E0604,
+  E0605,
+  E0606,
+  E0607,
+  E0608,
+  E0609,
+  E0610,
+  E0614,
+  E0615,
+  E0616,
+  E0617,
+  E0618,
+  E0619, // this error code is no longer emitted by the compiler
+  E0620,
+  E0621,
+  E0622,
+  E0623,
+  E0624,
+  E0625,
+  E0626,
+  E0627,
+  E0628,
+  E0631,
+  E0632, // this error code is no longer emitted by the compiler
+  E0633, // this error code is no longer emitted by the compiler
+  E0634,
+  E0635,
+  E0636,
+  E0637,
+  E0638,
+  E0639,
+  E0640,
+  E0641,
+  E0642,
+  E0643,
+  E0644,
+  E0646,
+  E0647,
+  E0648,
+  E0657,
+  E0658,
+  E0659,
+  E0660, // this error code is no longer emitted by the compiler
+  E0661, // this error code is no longer emitted by the compiler
+  E0662, // this error code is no longer emitted by the compiler
+  E0663, // this error code is no longer emitted by the compiler
+  E0664, // this error code is no longer emitted by the compiler
+  E0665, // this error code is no longer emitted by the compiler
+  E0666,
+  E0667,
+  E0668, // this error code is no longer emitted by the compiler
+  E0669, // this error code is no longer emitted by the compiler
+  E0670,
+  E0671, // this error code is no longer emitted by the compiler
+  E0687, // this error code is no longer emitted by the compiler
+  E0688, // this error code is no longer emitted by the compiler
+  E0689,
+  E0690,
+  E0691,
+  E0692,
+  E0693,
+  E0695,
+  E0696,
+  E0697,
+  E0698,
+  E0699,
+  E0700,
+  E0701,
+  E0703,
+  E0704,
+  E0705,
+  E0706,
+  E0708,
+  E0710,
+  E0711,
+  E0712,
+  E0713,
+  E0714,
+  E0715,
+  E0716,
+  E0717,
+  E0718,
+  E0719,
+  E0720,
+  E0722,
+  E0724,
+  E0725,
+  E0726,
+  E0727,
+  E0728,
+  E0729,
+  E0730,
+  E0731,
+  E0732,
+  E0733,
+  E0734,
+  E0735,
+  E0736,
+  E0737,
+  E0739,
+  E0740,
+  E0741,
+  E0742,
+  E0743,
+  E0744,
+  E0745,
+  E0746,
+  E0747,
+  E0748,
+  E0749,
+  E0750,
+  E0751,
+  E0752,
+  E0753,
+  E0754,
+  E0755,
+  E0756,
+  E0757,
+  E0758,
+  E0759, // this error code is no longer emitted by the compiler
+  E0760, // this error code is no longer emitted by the compiler
+  E0761,
+  E0762,
+  E0763,
+  E0764,
+  E0765,
+  E0766,
+  E0767,
+  E0768,
+  E0769,
+  E0770,
+  E0771,
+  E0772, // this error code is no longer emitted by the compiler
+  E0773,
+  E0774,
+  E0775,
+  E0776,
+  E0777,
+  E0778,
+  E0779,
+  E0780,
+  E0781,
+  E0782,
+  E0783,
+  E0784,
+  E0785,
+  E0786,
+  E0787,
+  E0788,
+  E0789,
+  E0790,
+  E0791,
+  E0792,
+  E0793,
+  E0794
+};
+
+// Custom hash function for ErrorCode, for older version of gcc
+namespace std {
+    template <>
+    struct hash<ErrorCode> {
+        size_t operator()(const ErrorCode& custom_hash) const {
+            return static_cast<size_t>(custom_hash);
+        }
+    };
+}
 
-  const char *m_str;
+static std::unordered_map<ErrorCode,const char *> ErrorCodeString = {
+  {ErrorCode::E0001, "E0001"},
```

Comment:
There is a slight chance of errors/mismatch in the future. It may be useful to create a macro like this:

```cpp
#define XSTR(A) str(A)
#define STR(A) #A
#define ERROR_CODE(NUM) E##NUM
#define TABLE_LINE(NUM) {ErrorCode::ERROR_CODE(NUM), xstr(ERROR_CODE(NUM))},
```

that way you'll be sure both number will always match as long as the macro is used everywhere in your table. If you do so, you should try to contain this macro at maximum, undef everything after usage.

Not entirely sure the added complexity is worth it though ? What do you think ?

Reply:
Yeah, I was thinking about, rustc was adding new error codes (like recently a new error code was added https://github.com/rust-lang/rust/commit/75563cd7253bb72876c5c97ff2f7813dd1485bc5 ), so we have to add new error codes in future. 

The way I generated all this stuff is by using the python script (can be found here: https://gist.github.com/MahadMuhammad/8c9d5fc88ea18d8c520937a8071d4185#file-rustc-errorcode-parser-py ), maybe this will also be helpful for future, so there are less chance for error in future. But adding this macro definitely reduces the chance of typo in the returned string and ensures that the proper `ErrorCodeString` is returned by the map. So, I think it's worth adding the macro.

Reply:
@P-E-P Done :)